### DISCUSSION TITLE
How to implement customProcessor in an autoinstrumented java service

### DISCUSSION THREAD

**Discussion by Raushan-Sapiens** [open-telemetry/opentelemetry-java-instrumentation#12749] [Q&A] [answered]
Hi,

We are using agent in our monolith application (non-spring) and would like to customize span with onStart method of Span Processor.

Agent version: **2.7.0**

Dependencies used are
```
            <dependency>
                <groupId>io.opentelemetry</groupId>
                <artifactId>opentelemetry-sdk</artifactId>
                <version>1.43.0</version>
            </dependency>
            <dependency>
                <groupId>io.opentelemetry</groupId>
                <artifactId>opentelemetry-sdk-trace</artifactId>
                <version>1.43.0</version>
            </dependency>
```
```
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;

public class CustomSpanProcessor implements SpanProcessor {
    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        // some logic
    }
```
While starting server we add our span registering logic
```
CustomSpanProcessor customSpanProcessor = new CustomSpanProcessor();
SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(customSpanProcessor)
                .build();
OpenTelemetrySdk.builder().setTracerProvider(tracerProvider).buildAndRegisterGlobal();
```

But still onStart method is not getting invoked. Please suggest

**Comment by steverao** [comment]
You can refer to [related demo](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSpanProcessor.java). Pay attention to override the method of `isStartRequired` and set its return to true.

**Comment by Raushan-Sapiens** [comment]
Hi @steverao , isStartRequired is to true and inspired by shared reference, but still it don't work

I think registering is a problem. I register it while we finish our start up activities of wildfly server.

```
CustomSpanProcessor customSpanProcessor = new CustomSpanProcessor();
SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(customSpanProcessor)
                .build();
OpenTelemetrySdk.builder().setTracerProvider(tracerProvider).buildAndRegisterGlobal();
```

**Reply by steverao** [reply]
If you register it manually, you can compare you codes with [related demo](https://github.com/open-telemetry/opentelemetry-java-examples/blob/main/sdk-usage/src/main/java/io/opentelemetry/sdk/example/ConfigureTraceExample.java). If you want to register it automatically, you can use `AutoConfigurationCustomizerProvider` like [demo](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoAutoConfigurationCustomizerProvider.java) here. If you can't solve it, can you provide a minimum demo to show how you use them?

**Reply by Raushan-Sapiens** [reply]
Hey @steverao ,

From manual process, I understand that when we need to modify spans then we need to fetch it from tracer the relevant span.

But we don't want to do this. We are instrumenting through agent and we need the customization from Demospanproceccore without actually fetching it from tracer. Should be done implicitly

**Reply by steverao** [reply, marked as answer]
> We are instrumenting through agent and we need the customization from Demospanproceccore without actually fetching it from tracer. Should be done implicitly

Take a look [related extension document](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/4d6b1cdf3831f16e01e4b4082803e56c2bb11033/examples/extension). You can add an extension to achieve your own DemoSpanProcessor when using OTel Java Agent.

**Reply by Raushan-Sapiens** [reply]
Thanks, I followed the doc and able to attach my agent
